File: src/main/java/problems/ValidParenthesis.java
package problems;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

// https://leetcode.com/problems/valid-parentheses/
public class ValidParenthesis {
  private static final Map<Character, Character> brackets = new HashMap<>();

  public ValidParenthesis() {
    brackets.put(')', '(');
    brackets.put(']', '[');
    brackets.put('}', '{');
  }

  public boolean isValid(String s) {
    final Stack<Character> charStack = new Stack<>();
    if (s.length() == 0) {
      return true;
    }
    for (char bracket : s.toCharArray()) {
      if (brackets.containsKey(bracket)) {
        if (charStack.isEmpty()) {
          return false;
        }
        char ch = charStack.pop();
        if (ch != brackets.get(bracket)) {
          return false;
        }
      } else {
        charStack.push(bracket);
      }
    }
    return charStack.isEmpty();
  }
}

File: python_problems/valid_parenthesis.py
brackets_dict = {
    ")": "(",
    "}": "{",
    "]": "["
}


def is_valid(input_expr) -> bool:
    if input_expr is None or len(input_expr) == 0:
        return False

    stack = []

    for i in input_expr:
        if i in brackets_dict:
            if len(stack) == 0:
                return False
            elif brackets_dict[i] != i:
                return False
        else:
            stack.append(i)

    return len(stack) == 0

